smack-experimental/src/main/java/org/jivesoftware/smackx/chat_markers/ChatMarkersManager.java

```diff
@@ -87,4 +204,101 @@ public boolean isSupportedByServer()
                 .serverSupportsFeature(ChatMarkersElements.NAMESPACE);
     }
 
+    /**
+     * Register a ChatMarkersListener. That listener will be informed about new
+     * incoming markable messages.
+     *
+     * @param listener ChatMarkersListener
+     * @return true, if the listener was not registered before
+     */
+    public boolean addIncomingChatMarkerMessageListener(ChatMarkersListener listener) {
+        synchronized (incomingListeners) {
+            return incomingListeners.add(listener);
+        }
+    }
+
+    /**
+     * Unregister a ChatMarkersListener.
+     *
+     * @param listener ChatMarkersListener
+     * @return true, if the listener was registered before
+     */
+    public boolean removeIncomingChatMarkerMessageListener(ChatMarkersListener listener) {
+        synchronized (incomingListeners) {
+            return incomingListeners.remove(listener);
+        }
+    }
+
+    /**
+     * From XEP-0333, Protocol Format: The Chat Marker MUST have an 'id' which is the 'id' of the
+     * message being marked.<br>
+     * In order to make Chat Markers works together with XEP-0085 as it said in
+     * 8.5 Interaction with Chat States, only messages with <tt>active</tt> chat
+     * state are accepted.
+     *
+     * @param message to be analyzed.
+     * @return true if the message contains a stanza Id.
+     * @see <a href="http://xmpp.org/extensions/xep-0333.html">XEP-0333: Chat Markers</a>
+     */
+    private boolean shouldDiscardMessage(Message message) {
+        if (StringUtils.isNullOrEmpty(message.getStanzaId())) {
+            return true;
+        }
+
+        if (CHAT_STATE_FILTER.accept(message)) {
```

Comment:
Although XEP-0333 (8.5 Interaction with Chat States) only say:
```
Chat Markers MAY be used alongside Chat States.
```
I added this after some test because I realize that without this I was marking `messages` like 'composing', 'paused' and 'active' that should not be marked.